src/v/cluster/client_quota_serde.h

```diff
@@ -0,0 +1,111 @@
+
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+#pragma once
+
+#include "base/seastarx.h"
+#include "seastar/core/sstring.hh"
+#include "serde/envelope.h"
+
+#include <vector>
+
+namespace cluster::client_quota {
+
+/// entity_key is used to key client quotas. It consists of multiple parts as a
+/// key can be a combination of key parts. Currently, only client id based and
+/// Redpanda-specific client id prefix-based quotas are supported, so all entity
+/// keys will consist of a single part, but in the future if we extend to client
+/// id and user principal based quotas, the entity key can contain two parts.
+struct entity_key
+  : serde::envelope<entity_key, serde::version<0>, serde::compat_version<0>> {
+    /// entity_type enumerates the different types of client quotas that we can
+    /// configure. The values of the variants are specified explicitly for serde
+    /// forwards/backwards compatibility.
+    enum class entity_type : uint8_t {
+        client_id_default = 0,
+        client_id = 1,
+        /// client_id_prefix is the Redpanda-specific client group quota type
+        client_id_prefix = 2,
+        // TODO: later we can extend to other Kafka-compatible entity types
+        // user_default = 3,
+        // user = 4,
+        // ip_default = 5,
+        // ip = 6,
+    };
+
+    struct part
+      : serde::envelope<part, serde::version<0>, serde::compat_version<0>> {
+        friend bool operator==(const part&, const part&) = default;
+        friend std::ostream& operator<<(std::ostream&, const part&);
+
+        template<typename H>
+        friend H AbslHashValue(H h, const part& e) {
+            return H::combine(std::move(h), e.type, e.value);
+        }
+        entity_type type;
+        ss::sstring value;
+        auto serde_fields() { return std::tie(type, value); }
+    };
+
+    auto serde_fields() { return std::tie(parts); }
+
+    friend bool operator==(const entity_key&, const entity_key&) = default;
+    friend std::ostream& operator<<(std::ostream&, const entity_key&);
+
+    template<typename H>
+    friend H AbslHashValue(H h, const entity_key& e) {
+        for (const auto& part : e.parts) {
+            h = AbslHashValue(std::move(h), part);
```

Comment:
Yes hashing the container directly works. I've turned the container of the `entity_key::part`'s into an `absl::flat_hash_set` instead of a vector, so that should give us the expected compound key equality/hashing behaviour.